langkit/envs.py

```diff
@@ -235,25 +226,46 @@ def _parse_actions(self, actions: List[EnvAction]) -> None:
         Analyze the given list of actions and extract pre/post actions, i.e.
         actions executed before and after handling children.
         """
+        def count(cls: Type[EnvAction],
+                  sequence: Optional[List[EnvAction]] = None) -> int:
+            """
+            Return the number of ``cls`` instances in ``sequence``, or in
+            ``actions`` if ``sequence`` is left to None.
+            """
+            if sequence is None:
+                sequence = actions
+            return len([a for a in sequence if isinstance(a, cls)])
+
         # If present, allow Do actions to come before SetInitialEnv
-        self.pre_initial_env_actions = []
+        first_actions = []
         if any(isinstance(a, SetInitialEnv) for a in actions):
             while actions and isinstance(actions[0], Do):
-                self.pre_initial_env_actions.append(actions.pop(0))
+                first_actions.append(actions.pop(0))
 
         # After that, allow one call to SetInitialEnv
         self.initial_env = None
         if actions and isinstance(actions[0], SetInitialEnv):
             self.initial_env = cast(SetInitialEnv, actions.pop(0))
+            first_actions.append(self.initial_env)
 
         pre, post = lsplit_by(
             lambda a: not isinstance(a, HandleChildren), actions
         )
 
+        for sequence in (pre, post):
```

Comment:
why not do the check on `actions` directly? (`count(SetInitialEnv, actions) == 0`)

Reply:
Ah yes, you’re right. `actions` is mutated, so it looks like actually all count checks could be done directly on it. I’ll simplify this part, thanks.